zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java

```diff
@@ -473,8 +474,23 @@ public Response runParagraph(@PathParam("notebookId") String notebookId,
       }
     }
 
-    note.run(paragraph.getId());
-    return new JsonResponse<>(Status.OK).build();
+    if (paragraph.getNoteReplLoader() == null) {
+      paragraph.setNoteReplLoader(note.getNoteReplLoader());
+    }
+
+    if (paragraph.getListener() == null) {
+      paragraph.setListener(note.getJobListenerFactory().getParagraphJobListener(note));
+    }
+
+    paragraph.run();
```

Comment:
paragraph supposed to not run in this way.
It should always run through [Scheduler](https://github.com/apache/incubator-zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/scheduler/Scheduler.java) that Interpreter provides.